src/main/java/org/springframework/data/hazelcast/repository/query/HazelcastPartTreeQuery.java

```diff
@@ -200,34 +200,27 @@ private Object executeFindQuery(final KeyValueQuery<?> query, final QueryMethod
 	private Object executePageSliceQuery(final Object[] parameters, final KeyValueQuery<?> query,
 			final QueryMethod queryMethod) {
 		long totalElements = -1;
-
 		int indexOfPageRequest = queryMethod.getParameters().getPageableIndex();
 		Pageable pageRequest = (Pageable) parameters[indexOfPageRequest];
 
-		/* TODO Eliminate count call for Slice, retrieve "rows+1" instead to determine if next page exists.
-		 */
-		if (query.getCritieria() == null) {
-			totalElements = this.keyValueOperations.count(queryMethod.getEntityInformation().getJavaType());
-		} else {
-			totalElements = this.keyValueOperations.count(query, queryMethod.getEntityInformation().getJavaType());
-		}
-
 		int requiredRows = pageRequest.getPageSize();
 
 		query.setOffset(pageRequest.getOffset());
-		query.setRows(pageRequest.getPageSize());
+		query.setRows(queryMethod.isSliceQuery() ? requiredRows + 1 : requiredRows);
 
 		Iterable<?> resultSet = this.keyValueOperations.find(query, queryMethod.getEntityInformation().getJavaType());
 		List<?> content = IterableConverter.toList(resultSet);
 
 		if (queryMethod.isPageQuery()) {
+			if (query.getCritieria() == null) {
+				totalElements = this.keyValueOperations.count(queryMethod.getEntityInformation().getJavaType());
+			} else {
+				totalElements = this.keyValueOperations.count(query, queryMethod.getEntityInformation().getJavaType());
+			}
 			return new PageImpl(content, pageRequest, totalElements);
 		} else {
-			boolean hasNext = totalElements > (query.getOffset() + query.getRows());
-			if (content.size() > requiredRows) {
-				content = content.subList(0, requiredRows);
-			}
-			return new SliceImpl(content, pageRequest, hasNext);
+			boolean hasNext = content.size() > requiredRows;
```

Comment:
`boolean hasNext = totalElements > (query.getOffset() + query.getRows());` was replaced by `boolean hasNext = content.size() > requiredRows;`
so `hasNext` boolean value of 3rd parameter of SliceImpl has modified logic. Is this intended behavior?

Reply:
As an optimization, I'm not querying `totalElements` of `SliceQuery` used. Instead, I used a trick to get requested rows + 1 entry as the query result. This is why that logic changed for SliceQueries. This is the actual change in this PR.